### PR TITLE
Issue #18711: Telemetry for credit card autofill

### PR DESCRIPTION
**Pull request by rocketsroger**
For #18711 

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by rocketsroger** [comment]
Needs https://github.com/mozilla-mobile/android-components/pull/10837

**Comment by rocketsroger** [comment]
# Request for data collection review form

All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.

1. What questions will you answer with this data?
  * This data will help us understand how users are interacting with credit cards and how they are using our autofill service.  
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? 
  * We are building out our services to include credit cards storage and autofill, and we need to know how useful this is for the users and how we can best serve them when it comes to their shopping experiences. This data will help us establish a baseline to make product decisions as we build out these services. 
3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
  * There are no other alternatives. 
4. Can current instrumentation answer these questions?
  * No. 
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki. 
Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.

Measurement Description | Data Collection Category | Tracking Bug #
-- | -- | --
Credit card saved: A counter of the number of credit cards that have been saved by the user.	| Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card deleted: A counter of the number of credit cards that have been deleted by the user. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card modified: The User has modified a credit card. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card form detected: A credit card form has been detected. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card autofilled: A credit card has been autofilled to a website. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card autofill prompt shown: A credit card autofill prompt has been shown. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card autofill prompt expanded: A credit card autofill prompt has been expanded. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card autofill prompt dismissed: A credit card autofill prompt has been dismissed. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card management add tapped: the add credit card button was tapped management in credit card management. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711
Credit card management card tapped: A credit card was tapped management in credit card management. | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/18711


6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
  * https://dictionary.telemetry.mozilla.org/apps/fenix 
7. How long will this data be collected? Choose one of the following:
  * One year (ending on September 1st, 2022), with the option to renew at that point. 
8. What populations will you measure?
  * All channels, all locales, all countries 
9. If this data collection is default on, what is the opt-out mechanism for users?
  * Default Glean SDK opt-out mechanism. 
10. Please provide a general description of how you will analyze this data.
  * Glean and Amplitude. 
11. Where do you intend to share the results of your analysis?
  *  Only on Glean, Amplitude, and with mobile teams. 
12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
   * No third-party tools.

**Comment by badboy** [comment]
Looping in @Dexterp37 for another set of Glean eyes. As I guided Roger through this it might be good to have someone not familiar yet review this briefly.

File: app/metrics.yaml [hunk 1]
```diff
@@ -5809,3 +5809,156 @@ recent_tabs:
...
+      - interaction
+    notification_emails:
+      - android-probes@mozilla.com
+    expires: "2022-09-01"
+  form_detected:
+    type: event
+    description: |
+      A credit card form was detected.
+    send_in_pings:
+      - metrics
```
Comment (Dexterp37) [inline app/metrics.yaml:5865]: Is this intentional? Events are usually sent in the [events](https://mozilla.github.io/glean/book/user/pings/events.html) ping. Note that you can also leave this property out for Glean to send the data automatically in the correct ping.

File: app/metrics.yaml [hunk 2]
```diff
@@ -5809,3 +5809,156 @@ recent_tabs:
...
+      - interaction
+    notification_emails:
+      - android-probes@mozilla.com
+    expires: "2022-09-01"
+  autofilled:
+    type: event
+    description: |
+      User has autofilled a credit card.
+    send_in_pings:
+      - metrics
```
Comment (Dexterp37) [inline app/metrics.yaml:5880]: As above: this applies to all the other event metrics added to this file.

File: app/metrics.yaml [hunk 3]
```diff
@@ -5809,3 +5809,156 @@ recent_tabs:
...
       - android-probes@mozilla.com
     expires: "2022-06-23"
+credit_cards:
+  saved:
+    type: counter
+    description: |
+      A counter of the number of credit cards that have been saved
+      manually by the user.
+    send_in_pings:
+      - metrics
```
Comment (Dexterp37) [inline app/metrics.yaml:5819]: Note that this property is not required, unless you specify a custom ping. This applies to the other definitions as well.

Comment (rocketsroger) [reply app/metrics.yaml:5819]: Understood.  I'll remove the `send_in_pings` property.  Thanks!

File: app/src/test/java/org/mozilla/fenix/components/metrics/MetricControllerTest.kt [hunk 4]
```diff
@@ -370,4 +371,105 @@ class MetricControllerTest {
...
+        )
+
+        event = controller.factToEvent(fact)
+        assertEquals(event, Event.CreditCardAutofillPromptDismissed)
+    }
+
+    @Test
+    fun `credit card events should be sent to enabled service`() {
+        val controller = ReleaseMetricController(
+            listOf(marketingService1),
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/test/java/org/mozilla/fenix/components/metrics/MetricControllerTest.kt]: Sure, I'll use the data service instead.  Thanks!

**Comment by travis79** [comment]
> # Request for data collection review form
> 
> All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.
> 
>     1. What questions will you answer with this data?
> 
> 
>     * This data will help us understand how users are interacting with credit cards and how they are using our autofill service.
> 
> 
>     1. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
>     * We are building out our services to include credit cards storage and autofill, and we need to know how useful this is for the users and how we can best serve them when it comes to their shopping experiences. This data will help us establish a baseline to make product decisions as we build out these services.
> 
> 
>     1. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
>     * There are no other alternatives.
> 
> 
>     1. Can current instrumentation answer these questions?
> 
> 
>     * No.
> 
> 
>     1. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.    
>        Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.
> 
> 
> Measurement Description 	Data Collection Category 	Tracking Bug #
> Credit card saved: A counter of the number of credit cards that have been saved by the user. 	Category 2 - interaction data 	#18711
> Credit card deleted: A counter of the number of credit cards that have been deleted by the user. 	Category 2 - interaction data 	#18711
> Credit card modified: The User has modified a credit card. 	Category 2 - interaction data 	#18711
> Credit card form detected: A credit card form has been detected. 	Category 2 - interaction data 	#18711
> Credit card autofilled: A credit card has been autofilled to a website. 	Category 2 - interaction data 	#18711
> Credit card autofill prompt shown: A credit card autofill prompt has been shown. 	Category 2 - interaction data 	#18711
> Credit card autofill prompt expanded: A credit card autofill prompt has been expanded. 	Category 2 - interaction data 	#18711
> Credit card autofill prompt dismissed: A credit card autofill prompt has been dismissed. 	Category 2 - interaction data 	#18711
> Credit card management add tapped: the add credit card button was tapped management in credit card management. 	Category 2 - interaction data 	#18711
> Credit card management card tapped: A credit card was tapped management in credit card management. 	Category 2 - interaction data 	#18711
> 
>     1. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
>     * [https://dictionary.telemetry.mozilla.org/apps/fenix   ](https://dictionary.telemetry.mozilla.org/apps/fenix%E2%80%A8)
> 
> 
>     1. How long will this data be collected? Choose one of the following:
> 
> 
>     * One year (ending on September 1st, 2022), with the option to renew at that point.
> 
> 
>     1. What populations will you measure?
> 
> 
>     * All channels, all locales, all countries
> 
> 
>     1. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
>     * Default Glean SDK opt-out mechanism.
> 
> 
>     1. Please provide a general description of how you will analyze this data.
> 
> 
>     * Glean and Amplitude.
> 
> 
>     1. Where do you intend to share the results of your analysis?
> 
> 
>     * Only on Glean, Amplitude, and with mobile teams.
> 
> 
>     1. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
> 
> 
>     * No third-party tools.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/apps/fenix )

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, through the "Send Usage Data" preference in the application settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

N/A, collection set to end or be renewed by 2022-09-01

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction data

5) Is the data collection request for default-on or default-off?

default-on

6) Does the instrumentation include the addition of **any *new* identifiers**?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

## Result

data-review+

**Review by eliserichards — APPROVED** [review]
Awesome! 👏 :shipit:

**Pull request merged**